From: Julian <juliannfairfax@protonmail.com>
Date: Thu, 2 Jan 2025 11:06:26 +0000
Subject: arm64: dts: sun50i-a64-pinetab: Add front/back cameras

PineTab has OV5640 back camera and GC2145 front camera. Add support for
 both.

Signed-off-by: =?utf-8?q?Ond=C5=99ej_Jirman?= <megi@xff.cz>
Signed-off-by: Ondrej Jirman <megi@xff.cz>
Signed-off-by: Andrey Skvortsov <andrej.skvortzov@gmail.com>
Signed-off-by: Julian <juliannfairfax@protonmail.com>
---
 .../boot/dts/allwinner/sun50i-a64-pinetab.dts      | 101 +++++++++++++++------
 1 file changed, 72 insertions(+), 29 deletions(-)

diff --git a/arch/arm64/boot/dts/allwinner/sun50i-a64-pinetab.dts b/arch/arm64/boot/dts/allwinner/sun50i-a64-pinetab.dts
index c5e7801..05194ac 100644
--- a/arch/arm64/boot/dts/allwinner/sun50i-a64-pinetab.dts
+++ b/arch/arm64/boot/dts/allwinner/sun50i-a64-pinetab.dts
@@ -11,6 +11,7 @@
 #include "sun50i-a64.dtsi"
 #include "sun50i-a64-cpu-opp.dtsi"
 
+#include <dt-bindings/media/video-interfaces.h>
 #include <dt-bindings/gpio/gpio.h>
 #include <dt-bindings/input/input.h>
 #include <dt-bindings/pwm/pwm.h>
@@ -60,40 +61,13 @@ hdmi_con_in: endpoint {
 		};
 	};
 
-	i2c-csi {
+	i2c_csi: i2c-csi {
 		compatible = "i2c-gpio";
 		sda-gpios = <&pio 4 13 GPIO_ACTIVE_HIGH>; /* PE13 */
 		scl-gpios = <&pio 4 12 GPIO_ACTIVE_HIGH>; /* PE12 */
 		i2c-gpio,delay-us = <5>;
 		#address-cells = <1>;
 		#size-cells = <0>;
-
-		/* Rear camera */
-		ov5640: camera@3c {
-			compatible = "ovti,ov5640";
-			reg = <0x3c>;
-			pinctrl-names = "default";
-			pinctrl-0 = <&csi_mclk_pin>;
-			clocks = <&ccu CLK_CSI_MCLK>;
-			clock-names = "xclk";
-
-			AVDD-supply = <&reg_dldo3>;
-			DOVDD-supply = <&reg_aldo1>;
-			DVDD-supply = <&reg_eldo3>;
-			reset-gpios = <&pio 4 14 GPIO_ACTIVE_LOW>; /* PE14 */
-			powerdown-gpios = <&pio 4 15 GPIO_ACTIVE_HIGH>; /* PE15 */
-
-			port {
-				ov5640_ep: endpoint {
-					remote-endpoint = <&csi_ep>;
-					bus-width = <8>;
-					hsync-active = <1>; /* Active high */
-					vsync-active = <0>; /* Active low */
-					data-active = <1>;  /* Active high */
-					pclk-sample = <1>;  /* Rising */
-				};
-			};
-		};
 	};
 
 	speaker_amp: audio-amplifier {
@@ -144,10 +118,15 @@ &cpu3 {
 };
 
 &csi {
+	pinctrl-0 = <&csi_pins>;
 	status = "okay";
 
 	port {
-		csi_ep: endpoint {
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		csi_ov5640_ep: endpoint@0 {
+			reg = <0>;
 			remote-endpoint = <&ov5640_ep>;
 			bus-width = <8>;
 			hsync-active = <1>; /* Active high */
@@ -155,6 +134,16 @@ csi_ep: endpoint {
 			data-active = <1>;  /* Active high */
 			pclk-sample = <1>;  /* Rising */
 		};
+
+		csi_gc2145_ep: endpoint@1 {
+			reg = <1>;
+			remote-endpoint = <&gc2145_ep>;
+			bus-width = <8>;
+			hsync-active = <1>;
+			vsync-active = <1>;
+			data-active = <1>;
+			pclk-sample = <1>;
+		};
 	};
 };
 
@@ -193,6 +182,60 @@ &ehci1 {
 	status = "okay";
 };
 
+&i2c_csi {
+	pinctrl-names = "default";
+	pinctrl-0 = <&csi_mclk_pin>;
+
+	gc2145: front-camera@3c {
+		compatible = "galaxycore,gc2145";
+		reg = <0x3c>;
+		clocks = <&ccu CLK_CSI_MCLK>;
+		avdd-supply = <&reg_dldo3>;
+		dvdd-supply = <&reg_aldo1>;
+		iovdd-supply = <&reg_eldo3>;
+		reset-gpios = <&pio 4 16 (GPIO_ACTIVE_LOW | GPIO_OPEN_DRAIN)>; /* PE16 */
+		powerdown-gpios = <&pio 4 17 (GPIO_ACTIVE_HIGH | GPIO_OPEN_DRAIN)>; /* PE17 */
+		rotation = <90>;
+		orientation = <0>;
+
+		port {
+			gc2145_ep: endpoint {
+				remote-endpoint = <&csi_gc2145_ep>;
+				bus-type = <MEDIA_BUS_TYPE_PARALLEL>;
+				bus-width = <8>;
+				hsync-active = <1>;
+				vsync-active = <1>;
+				pclk-sample = <1>;
+			};
+		};
+	};
+
+	ov5640: rear-camera@4c {
+		compatible = "ovti,ov5640";
+		reg = <0x4c>;
+		clocks = <&ccu CLK_CSI_MCLK>;
+		clock-names = "xclk";
+		AVDD-supply = <&reg_dldo3>;
+		DOVDD-supply = <&reg_aldo1>; /* shared with AFVCC */
+		DVDD-supply = <&reg_eldo3>;
+		reset-gpios = <&pio 4 14 (GPIO_ACTIVE_LOW | GPIO_OPEN_DRAIN)>; /* PE14 */
+		powerdown-gpios = <&pio 4 15 (GPIO_ACTIVE_HIGH | GPIO_OPEN_DRAIN)>; /* PE15 */
+		rotation = <270>;
+		orientation = <1>;
+
+		port {
+			ov5640_ep: endpoint {
+				remote-endpoint = <&csi_ov5640_ep>;
+				bus-width = <8>;
+				hsync-active = <1>; /* Active high */
+				vsync-active = <0>; /* Active low */
+				data-active = <1>;  /* Active high */
+				pclk-sample = <1>;  /* Rising */
+			};
+		};
+	};
+};
+
 &i2c0 {
 	status = "okay";
 
